File: java/lesson 10/HowSumBottomUp.java
import java.util.*;

class Solution {

    public List<Integer> howSum(int targetSum, int[] numbers) {
        Integer[] dp = new Integer[targetSum + 1];
        dp[0] = 0;

        for (int i = 0; i <= targetSum; i++) {
            if (dp[i] == null) continue;            
            for (int number : numbers) {
    		    if (i + number > targetSum) continue;
                dp[i + number] = number;
                if (i + number == targetSum) break;
            }
        }
        
        List<Integer> result = new ArrayList<>();
        int current = targetSum;
        while (dp[current] != null && current != 0) {
            result.add(dp[current]);
            current = current - dp[current];
        }
        return result;
    }
}

File: csharp/lesson 10/HowSumBottomUp.cs
class Solution
{
    public List<int> HowSum(int targetSum, int[] numbers)
    {
        var dp = new int?[targetSum + 1];
        dp[0] = 0;

        for (int i = 0; i <= targetSum; i++)
        {
            if (dp[i] == null) continue;            
            foreach (var number in numbers)
            {
                if (i + number > targetSum) continue;
                dp[i + number] = number;
                if (i + number == targetSum) break;
            }
        }

        var result = new List<int>();
        var current = targetSum;
        while (dp[current] != null && current != 0)
        {
            result.Add(dp[current].Value);
            current = current - dp[current].Value;
        }
        return result;
    }
}
